### DISCUSSION TITLE
Start express js server when opening tauri app?

### DISCUSSION THREAD

**Discussion by grey4owl** [tauri-apps/tauri#2613] [Q&A] [answered]
Hi everyone! 😃 I would like to use express js with tauri app, but i can't figure out how to run express server and tauri app at the same time? 🤔  I tried to call express js in the main file where all the code for my application is but before that code..

**`main.js`**
```javascript
import { appWindow } from "@tauri-apps/api/window";
import { invoke } from "@tauri-apps/api/tauri";

import express from "express";

const app = express();
const port = 3000;
 
app.listen(3000, () =>
  console.log('App listening on port 3000!'),
);

// below is the rest of the code for tauri app...
```

I'm using rollup.js  module bundler and this is configuration that I use:

**`rollup.config.js`**
```
import { nodeResolve } from '@rollup/plugin-node-resolve';

export default {
  input: 'main.js',
  output: {
    sourcemap: 'inline',
    format: 'iife',
    name: 'app',
    file: 'client/js/bundle.js',
    globals: {
      '@tauri-apps/api/window': 'appWindow',
      '@tauri-apps/api/tauri': 'invoke',
      'express': 'express'
    }
  },
  external: ['@tauri-apps/api', 'express'],
  plugins: [nodeResolve({
    moduleDirectories: ['node_modules']
  })]
};
```

When bundling, rollup does not show any problem, but when I run the application `npm run tauri dev` in the console it shows that `express` is not defined. Is there some better solution to run express and tauri together?

**Comment by amrbashir** [comment, marked as answer]
ExpressJS is an NodeJS library, it won't work  in Tauri's webview. 

If you need the functionality of a webserver, then either use the sidecar pattern or look for an equivalent rust crate of ExpressJS.

**Reply by grey4owl** [reply]
So the only option is to use something like [rocket.rs](https://rocket.rs) ? Or using the sidecar option to insert express js?

**Reply by amrbashir** [reply]
It is not the only option but the recommended one, other option is sidecar.

I don't know your use-case for a webserver inside a Tauri app you probably don't need it.

**Reply by grey4owl** [reply]
I am trying to make a CRUD application that has a login system and database..

**Reply by amrbashir** [reply]
That seems like you just need to deploy a standalone webserver where you are gonna host your users database, no need to embed it in Tauri.

**Reply by grey4owl** [reply]
Yes but how do i connect the mysql database to the application if i don't have something like express js to handle requests? I wanted to connect the mysql database hosted online with the interface in the application. I am a beginner with tauri so i am wondering if it is possible to do that without node js? 🥲

**Reply by amrbashir** [reply]
If you have an online database, then there is a good chance it already has an api endpoints that you can use, you wouldn't need to build the api yourself, but if it doesn't have an existing api, then you need to build one with ExpressJS and deploy it. All of this is out of Tauri scope.

After you build your database and its api, you can use the api through http requests from Tauri app.

**Comment by code-briomar** [comment]
I may have a way to work around it.  ( You'll be able to do what you initially wanted - **run both the Tauri app and Express app concurrently** ) 

I have placed all the **Express code** in a separate folder  **inside** the tauri project.

<img width="126" alt="image" src="https://github.com/tauri-apps/tauri/assets/82029545/f7242fef-90e4-4bd6-8626-77b6c3f4cdd0">

I then modified the **package.json** file in my tauri app, as such : 


<img width="361" alt="image" src="https://github.com/tauri-apps/tauri/assets/82029545/68d08266-2c4f-48bf-8ab3-76b8ea03007e">


So when I run `yarn start` on my console it launches both the Express app with my API endpoints and my tauri app which run concurrently.

**Reply by phantomdenkali** [reply]
this seems interesting, will give this a try. Does this effectively allow the server to be packaged with the same tauri "client" app

**Reply by code-briomar** [reply]
Yes it does. you can also use the **pm2** package so that your Express app listens to changes in your code and you won't have to restart your express app every time you add changes to your code. 

[sanCodeTauri](https://github.com/sanCode-2-0/sanCodeTauri)

**Reply by phantomdenkali** [reply]
thanks for the reply. how are you handling auto-updates for the express app and retaining user state?

**Reply by code-briomar** [reply]
When starting the express app using pm2, I use a --watch flag. It looks like this : `pm2 start myExpressApp.js --watch` .  This allows pm2 to monitor changes in the file and automatically restart the app when a change is detected.

Concerning retaining user state, I'm currently not using that in the sanCodeTauri app, but normally you would opt for services like firebase to help you with that.

**Reply by loprima-l** [reply]
Won't the user have to download node also ?

**Reply by phantomdenkali** [reply]
I was thinking of using a packager like esbuild to package all dependencies
into a single binary.
https://dev.to/midnqp/bundling-nodejs-into-single-executable-binary-l3g

I haven't gotten around to doing any of this yet...so very theoretical at
this point.

**Reply by code-briomar** [reply]
@loprima-l  Yes

**Reply by loprima-l** [reply]
That sound like a good idea. But I'm not sure to understand how it can work without having node as the integrated js thing doesn't have access to OS things.

**Comment by code-briomar** [comment]
@phantomdenkali I don't find it necessary to package all dependancies.

**Reply by loprima-l** [reply]
I don't think this would help us to run it without installing node right ?

**Reply by code-briomar** [reply]
You're right.

**Comment by loprima-l** [comment]
I think the solution could be using NEXE and compiling the backend in another app, and using it like a real server, it would help also in the case you want multiple TAURI client to interact together.

**Comment by code-briomar** [comment]
@loprima-l Here's how I'm managing them in a single Tauri app


https://github.com/tauri-apps/tauri/assets/82029545/f0750e4e-15ac-4a8f-8a52-330233c42da5

**Reply by loprima-l** [reply]
Ok but we still got to install Node JS right ?

**Reply by code-briomar** [reply]
Yeah

**Reply by sakaria26** [reply]
How did you manage to get this working in production?

**Reply by gp3098** [reply]
我们不可能让最终客户安装nodejs，这和单独启动nodejs没有区别。既然用rust了，那么我认为rust也可以安装数据库执行curd.你可以说你对rust不熟悉，但是这不是改用nodejs的理由

**Reply by zzx58-027** [reply]
It wont be working in production. If I am wrong, please help pointing out.
